### PR TITLE
Support role and group permissions in file sources plugins

### PR DESCRIPTION
**Pull request by davelopez**
## What did you do? 
- Added new security options to file sources plugins:
  ````YAML
    requires_roles: some_role
    requires_groups: some_group
  ````
- Included unit tests to check the security rules defined in the options are enforced.
- Included some integration tests to check user access using the `posix` file source plugin.
- Fixed a small inconsistency between the legacy and the new FastAPI `remote_files` API default responses. This was detected while writing the tests. See this commit https://github.com/galaxyproject/galaxy/commit/4235e6e675cec92fc783096a3949bc9369d107e0

#### TODO
- [x] Use basic boolean logic to define the rules to match groups and roles.
- [x] Create unit tests for the boolean logic parser.

## Why did you make this change?
The details are in #11768

## How to test the changes? 
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

### CODE REVIEW

**Comment by davelopez** [comment]
I've added a new generic [BooleanExpressionEvaluator](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063) class (using `pyparsing` which was already in the dependencies) that can parse boolean expressions using `AND`, `OR` and `NOT` operators.

The way of evaluating each *token* is by providing a custom implementation of the [TokenEvaluator interface](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063#diff-473c28c74e0aa25d6e54c9d16260f9e440293002b339d7f2fd4b1b96a624a284R41) that determines if a token should be evaluated to `True` or `False` depending on your custom logic. For the case of checking the groups and roles of a user, I have implemented and used the *somewhat* generic [TokenContainerEvaluator](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063#diff-473c28c74e0aa25d6e54c9d16260f9e440293002b339d7f2fd4b1b96a624a284R154) and use as the input set the names of the groups or roles of the user. Other implementations of the `TokenEvaluator` can be used to deal with other cases that may need this kind of expression parser.

Another important thing to mention is that the `BooleanExpressionEvaluator` can parse only tokens with a valid format but this format can be personalized using the [`token_format` parameter](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063#diff-473c28c74e0aa25d6e54c9d16260f9e440293002b339d7f2fd4b1b96a624a284R122). Even so, whitespace characters are not allowed and, right now, groups and role names can be any arbitrary string including whitespace characters which may break the parsing. Maybe is a good idea to limit the input name of new groups or roles to something a bit more restricted and use the description field to go wild 😄

I think this is ready to review :)

**Comment by davelopez** [comment]
The nice catch by the [failing integration test](https://github.com/galaxyproject/galaxy/runs/2254711212#step:12:1296) made me wonder if the best way to deal with a `ParseException` during `evaluate_expression` is to return `False` instead of raising the exception after [logging it](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063#diff-473c28c74e0aa25d6e54c9d16260f9e440293002b339d7f2fd4b1b96a624a284R151) or add another try/catch in the upper layer [here](https://github.com/galaxyproject/galaxy/pull/11769/commits/e034808fd4b1b9bf2d7a7a0d67633a5e324e7647#diff-644fe18408381b506c4abc27ffe07d42b8a8dc55a5a833f705662f65fe60c6d6R192) to always return `False` on general exceptions 🤔

**Comment by bgruening** [comment]
> The nice catch by the [failing integration test](https://github.com/galaxyproject/galaxy/runs/2254711212#step:12:1296) made me wonder if the best way to deal with a `ParseException` during `evaluate_expression` is to return `False` instead of raising the exception after [logging it](https://github.com/galaxyproject/galaxy/pull/11769/commits/9dd0d66bd42f4fbd7a364669cea9ad5eafb33063#diff-473c28c74e0aa25d6e54c9d16260f9e440293002b339d7f2fd4b1b96a624a284R151) or add another try/catch in the upper layer [here](https://github.com/galaxyproject/galaxy/pull/11769/commits/e034808fd4b1b9bf2d7a7a0d67633a5e324e7647#diff-644fe18408381b506c4abc27ffe07d42b8a8dc55a5a833f705662f65fe60c6d6R192) to always return `False` on general exceptions thinking

I can not answer that, maybe @jmchilton as an opinion here, but we should make sure that only this one file-source plugin gets skipped completely (its better to know how a data resource, than offering it with wrong permissions) and not all others or even worse than all uploads are affected. Like here: https://github.com/galaxyproject/galaxy/issues/11766 From that standpoint a try/except in the higher layer sounds interesting.

**Comment by jmchilton** [comment]
1. If it is possible to evaluate the expressions during parsing and just stop Galaxy from starting if there is a bad config - I think that is the ideal way to tell admins about config problems. It is a little extra work though - I'm not sure if the interface makes that easy?

2. I think pushing exceptions to higher layers here is generally right? I think @bgruening is also right though that it is more important to ensure we're not offering file sources with the incorrect permissions.

**Comment by davelopez** [comment]
Many thanks for the suggestions!
I have added a [_validate_security_rules()](https://github.com/galaxyproject/galaxy/pull/11769/commits/b1ac78a19b28d729ce68a275e89a6dfd87d774d8#diff-644fe18408381b506c4abc27ffe07d42b8a8dc55a5a833f705662f65fe60c6d6R104) method when parsing the file source plugin configuration that will raise a `ConfigurationError` exception if any of the boolean expressions have either syntax or grammatical errors.
If we set the security rule to something not valid like:
````YAML
  requires_roles: this is not valid
````
The error will be similar to this:
````
galaxy.util.bool_expressions ERROR 2021-04-04 14:45:08,557 [pN:MainProcess,p:7519,tN:MainThread] BooleanExpressionEvaluator unable to evaluate expression => this is not valid
Traceback (most recent call last):
  File "lib/galaxy/util/bool_expressions.py", line 147, in evaluate_expression
    res = self.boolExpr.parseString(expr, parseAll=True)[0]
  File "/home/davelopez/dev/galaxy/.venv/lib/python3.7/site-packages/pyparsing.py", line 1955, in parseString
    raise exc
  File "/home/davelopez/dev/galaxy/.venv/lib/python3.7/site-packages/pyparsing.py", line 3814, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected end of text, found 'i'  (at char 5), (line:1, col:6)
Traceback (most recent call last):
  File "lib/galaxy/webapps/galaxy/buildapp.py", line 60, in app_pair
    app = galaxy.app.UniverseApplication(global_conf=global_conf, **kwargs)
  File "lib/galaxy/app.py", line 197, in __init__
    super().__init__(fsmon=True, **kwargs)
  File "lib/galaxy/app.py", line 172, in __init__
    self.file_sources = self._register_singleton(ConfiguredFileSources, ConfiguredFileSources.from_app_config(self.config))
  File "lib/galaxy/files/__init__.py", line 165, in from_app_config
    return ConfiguredFileSources(file_sources_config, config_file, load_stock_plugins=True)
  File "lib/galaxy/files/__init__.py", line 26, in __init__
    file_sources = self._load_plugins_from_file(conf_file)
  File "lib/galaxy/files/__init__.py", line 60, in _load_plugins_from_file
    return self._parse_plugin_source(plugin_source)
  File "lib/galaxy/files/__init__.py", line 74, in _parse_plugin_source
    dict_to_list_key="id",
  File "lib/galaxy/util/plugin_config.py", line 39, in load_plugins
    dict_to_list_key=dict_to_list_key,
  File "lib/galaxy/util/plugin_config.py", line 93, in __load_plugins_from_dicts
    plugin = plugins_dict[plugin_type](**plugin_kwds)
  File "lib/galaxy/files/sources/posix.py", line 28, in __init__
    props = self._parse_common_config_opts(kwd)
  File "lib/galaxy/files/sources/__init__.py", line 104, in _parse_common_config_opts
    self._validate_security_rules()
  File "lib/galaxy/files/sources/__init__.py", line 206, in _validate_security_rules
    raise ConfigurationError(_get_error_msg_for("requires_roles"))
galaxy.exceptions.ConfigurationError: Invalid boolean expression for 'requires_roles' in Posix file source plugin configuration.
````

File: test/unit/util/test_bool_expressions.py [hunk 1]
```diff
@@ -14,6 +14,27 @@
...
+    ("not T3 or (T3 AND token_2)", True),
+    ("T1 and (T3 OR token_2)", True),
+    ("(T3 OR T1) and not (T3 OR valid_token)", True),
+]
+
+INVALID_EXPRESSIONS_TESTS = [
+    "",
+    "23 45",
+    "'some quoted str' and not T1",
+    "invalid expression",
```
Comment (bgruening) [inline test/unit/util/test_bool_expressions.py:35]: can we add a few more here like

```python
"T1 and and T2",
"T1 not and T3",
"(T1 and T3",
"T1 or T3)",
"T1 and or T2",
```

Comment (davelopez) [reply test/unit/util/test_bool_expressions.py:35]: Done! :)

**Comment by davelopez** [comment]
I think the failing selenium test is unrelated. Running in local seems to pass.

**Comment by nsoranzo** [comment]
> I think the failing selenium test is unrelated. Running in local seems to pass.

Very likely! I've seen several SQLite-related test failures in Selenium tests, I am not sure why the PostgreSQL database is used for `database_connection` but not automatically reused for `amqp_internal_connection` as it should be. From https://github.com/galaxyproject/galaxy/pull/11769/checks?check_run_id=2276839926 :

```
test_driver INFO 2021-04-06 09:39:23,243 [pN:main,p:4164,tN:MainThread] Galaxy database connection: ***localhost:5432/galaxy?client_encoding=utf8
...
galaxy.queue_worker INFO 2021-04-06 09:39:23,632 [pN:main,p:4164,tN:MainThread] Initializing main Galaxy Queue Worker on sqlalchemy+sqlite:////tmp/tmph7ytzdfv/tmp4kviu6eb/tmpnu5fs95w/database/control.sqlite?isolation_level=IMMEDIATE
```


@ic4f maybe you have a clue?

**Comment by jmchilton** [comment]
This is really nicely done, thanks so much!

**Comment by nsoranzo** [comment]
> I've seen several SQLite-related test failures in Selenium tests, I am not sure why the PostgreSQL database is used for database_connection but not automatically reused for amqp_internal_connection as it should be.

It looks like this is on purpose: https://github.com/galaxyproject/galaxy/pull/8469 . Not sure if we still need this workaround though? @jmchilton

**Pull request merged**